src/main/java/org/stroganov/ws/handlers/UserValidator.java

```diff
@@ -0,0 +1,104 @@
+package org.stroganov.ws.handlers;
+
+import jakarta.xml.soap.*;
+import jakarta.xml.ws.handler.MessageContext;
+import jakarta.xml.ws.handler.soap.SOAPHandler;
+import jakarta.xml.ws.handler.soap.SOAPMessageContext;
+import jakarta.xml.ws.soap.SOAPFaultException;
+import org.apache.log4j.Logger;
+import org.stroganov.dao.LibraryDAO;
+import org.stroganov.entities.User;
+import org.stroganov.util.DataManager;
+
+import javax.xml.namespace.QName;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+
+public class UserValidator implements SOAPHandler<SOAPMessageContext> {
+    public static final String CREDENTIALS_ERROR = "Не удалось получить правильный пароль для пользователя:";
+    private final Logger logger = Logger.getLogger(UserValidator.class);
+    protected final LibraryDAO libraryDAO = DataManager.getLibraryDAO();
+    protected User user = null;
+
+    @Override
+    public boolean handleMessage(SOAPMessageContext context) {
+        boolean isRequest = (Boolean) context.get(MessageContext.MESSAGE_OUTBOUND_PROPERTY);
+        if (!isRequest) {
+            try {
+                SOAPMessage soapMsg = context.getMessage();
+                SOAPEnvelope soapEnv = soapMsg.getSOAPPart().getEnvelope();
+                SOAPHeader soapHeader = soapEnv.getHeader();
+                if (soapHeader == null) {
+                    soapHeader = soapEnv.addHeader();
+                    generateSOAPErrMessage(soapMsg, "No SOAP header.");
+                }
+                Iterator<SOAPHeaderElement> it = soapHeader.extractHeaderElements(SOAPConstants.URI_SOAP_ACTOR_NEXT);
+                if (it == null || !it.hasNext()) {
+                    generateSOAPErrMessage(soapMsg, "No header block for next actor.");
+                }
+                String userLogin = null;
+                String password = null;
+                assert it != null;
+                while (it.hasNext()) {
+                    Node node = it.next();
+                    if (node.getNodeName().equals("login")) {
+                        userLogin = node.getValue();
+                    }
+                    if (node.getNodeName().equals("password")) {
+                        password = node.getValue();
+                    }
+                }
+                if (userLogin == null || password == null) {
+                    generateSOAPErrMessage(soapMsg, "No Password");
+                    return false;
+                }
+                if (checkUser(userLogin, password)) {
```

Comment:
this code duplicate in AdminValidator
extract to fun 
```suggestion
    protected User getUserFromHeaders(Iterator<SOAPHeaderElement> headers) {
    //your code
     return libraryDAO.findUser(userLogin);
    }
```

Reply:
Fixed. May be this part: 
                SOAPMessage soapMsg = context.getMessage();
                SOAPEnvelope soapEnv = soapMsg.getSOAPPart().getEnvelope();
                SOAPHeader soapHeader = soapEnv.getHeader();

                if (soapHeader == null) {
                    soapHeader = soapEnv.addHeader();
                    generateSOAPErrMessage(soapMsg, "No SOAP header.");
                }
could be included in 'getAuthorizedUserFromHeaders' method too ?

Reply:
yeah, but let's create a new fuction for it. getAuthorizedUserFromHeaders is too huge 
```  suggested 
        
private Iterator<SOAPHeaderElement> getSoapHeaders(SOAPMessageContext context) {      
                SOAPMessage soapMsg = context.getMessage();
                SOAPEnvelope soapEnv = soapMsg.getSOAPPart().getEnvelope();
                SOAPHeader soapHeader = soapEnv.getHeader();
                if (soapHeader == null) {
                    soapHeader = soapEnv.addHeader();
                    generateSOAPErrMessage(soapMsg, "No SOAP header.");
                }
                return soapHeader.extractHeaderElements(SOAPConstants.URI_SOAP_ACTOR_NEXT);
}
```

Reply:
Done. c5ae7228eae0b53b7942375b24c1464a66b23da5